Conditional dependencies in lockdir

gridbugs opened the pull request:
This changes lockdir generation so that dependencies which are conditional on the variables `:with-test` and `:with-doc` are added to the lockdir.

gridbugs committed:
Move opam solver contexts to separate file

The context modules can be quite verbose and it adds a lot of noise to
the opam module, so moving them to their own file to improve
readibility.

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs committed:
Solver includes test and doc deps of local packages

This adds packages depended on by local projects with filters :with-test
and :with-doc to the lockdir.

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

Leonidas-from-XIV commented on src/dune_pkg/opam_solver.ml:
```diff
@@ -0,0 +1,190 @@
+  (* Override [filter_deps] so that local packages get test and doc dependencies *)
+  let filter_deps t =
+    let local_packages = Base_context.local_packages t.base_context in
+    filter_deps_env ~local_packages ~env:(fun pkg v ->
```
Isn't this the env function from the solver? Maybe it can be reused?

gridbugs replied:
Actually yes, I've since dug a bit deeper into opam filters in another change I'm working on and I think we can simplify this by resolving the `test` and `doc` variables to true before calling the base context's `filter_deps` on the result. Will update.

Leonidas-from-XIV commented on src/dune_pkg/opam_solver.ml:
```diff
@@ -0,0 +1,190 @@
+        else
+          let r = OpamPackageVar.resolve_switch ~package:pkg t.switch_state v in
+          if r = None then
+            OpamConsole.warning "Unknown variable %S"
```
It should probably not warn to the Opam console.

gridbugs replied:
Oops yes you're right.

Leonidas-from-XIV reviewed (commented):
Looks good and a very useful change. Some nitpicking on the code from the 0install solver.

gridbugs commented on src/dune_pkg/opam_solver.ml:
```diff
@@ -0,0 +1,190 @@
+  in
+  filtered_formula
+  |> OpamFilter.partial_filter_formula (env pkg)
+  |> OpamFilter.filter_deps ~build:true ~post:true ~test:pkg_is_local
```
It means that dependencies marked as "post" will be resolved. I've never encountered this variable in the wild but it tells opam that the dependency can be compiled after the current package. https://opam.ocaml.org/doc/Manual.html#pkgvar-post I set it to true because that's what `Opam_0install.Dir_context` does and it seems like the most useful default setting. Perhaps one day we'll need to add support to dune to set it to false to deal with dependency cycles.

Leonidas-from-XIV replied:
Thanks for the explanation, I have never encountered these before. Out of curiosity I checked opam-repository and there are two non-compiler packages (only 2 versions overall) that use these: `sail.0.15` and `ctypes-foreign.0.18.0` so indeed these don't seem to be used much.

gridbugs commented:
I removed all the code borrowed/modified from opam_0install and instead added a module `Filtered_formula` with functions for transforming opam formulas and now `filter_deps` just uses that to replace `with-test` and `with-doc` with `true` before calling the base context's `filter_deps` on the result. This also means we could get rid of some of the `*_for_dune` modules previously in this change.

Also in the process I noticed that we were setting the `with-dev-setup` variable which I don't think we want so now we just set the `with-test` and `with-doc` variables to true.

Leonidas-from-XIV commented on src/dune_pkg/opam_solver.ml:
```diff
@@ -0,0 +1,131 @@
+        | Constraint _ as constraint_ -> `Formula (Atom constraint_)))
+
+  (* update a filtered formula replacing instances of a given variable with a given boolean value *)
+  let resolve_simple_bool_variable ~variable_name ~value =
```
Maybe `variable_name` would be better as an `OpamVariable`, then one could use its `equal` function, thus making the named arguments less important.

Leonidas-from-XIV commented on src/dune_pkg/opam_solver.ml:
```diff
@@ -0,0 +1,131 @@
+  (* [map_formula_filters_bottom_up t ~f] calls [f] on each node of each filter
+     in t, bottom-up *)
+  let map_filters_bottom_up ~f =
+    OpamFilter.gen_filter_formula
```
Cool, didn't know these functions. `map_filters_bottom_up` seems quite useful.

gridbugs replied:
Yeh `OpamFilter.map_up` is very handy. I just learnt about it recently too.

rgrinberg commented:
A couple of things that are missing from this PR:

* The ability to customize this behavior. In particular, it should be possible not to have docs or test deps. Also, it should be possible to include one of the two.
* The second thing we're missing is a way to propagate which deps were added because of the deps or tests flags. This is so that installing the software can use the same lock file (when applicable).

The first one needs to be done in this PR, while the second one can be done in subsequent work.

gridbugs commented:
>     * The second thing we're missing is a way to propagate which deps were added because of the deps or tests flags. This is so that installing the software can use the same lock file (when applicable).

Can you clarify this point. Do you mean including in package metadata whether the package was included in the lockdir because it's a test/doc dependency so that we can omit those packages when using the lockdir for installation?

gridbugs commented:
Cool makes sense. I'll update this PR so that `with-test` and `with-doc` can be configured per build context and make a new PR which tracks which packages were included due to these flags in a later change.

gridbugs committed:
Add solver env for keeping track of solver flags

The solver env is configurable per build context and contains variable
assignments used by the solver when solving dependencies. Currently it
just contains the flags with-test and with-doc but could easily be
extended to track system variables. Its purpose currently overlaps with
that of `Opam.Env.t` but the plan is to fold that into `Solver_env.t` in
a future change.

Signed-off-by: Stephen Sherratt <stephen@sherra.tt>

gridbugs commented:
@rgrinberg I've added a `Solver_env.t` type which stores variables used by the solver. Currently it just stores the flags `:with-test` and `:with-doc` but I think it would make sense to use this for all opam variables (e.g. system variables) and replace the `Opam.Env.t` type completely. I'm experimenting with that on another branch but I won't add it to this PR so it doesn't get too complicated.

rgrinberg commented on src/dune_pkg/opam_solver.ml:
```diff
@@ -0,0 +1,136 @@
+
+  let with_flags flags filtered_formula =
+    Solver_env.Flag.Set.to_list flags
+    |> List.fold_left ~init:filtered_formula ~f:(fun acc flag ->
```
You could just add a fold function to `Solver_env.Flag.Set`. It doesn't really matter much, but why favor list conversion when all you need is a fold?

rgrinberg commented on src/dune_pkg/opam_solver.mli:
```diff
@@ -0,0 +1,20 @@
+       solver_env:Solver_env.t
+    -> switch_state:OpamStateTypes.unlocked OpamStateTypes.switch_state
+    -> local_packages:OpamFile.OPAM.t OpamTypes.name_map
+    -> prefer_oldest:bool
```
Don't we have a version preference type somewhere that we can use?

rgrinberg commented on src/dune_pkg/opam_solver.mli:
```diff
@@ -0,0 +1,20 @@
+
+  val create_dir_context :
+       solver_env:Solver_env.t
+    -> env:(string -> OpamVariable.variable_contents option)
```
Not entirely related, but I would prefer this naked `string` type turned into something abstract.

gridbugs replied:
I'm planning to remove this argument soon when I move all environment variables inside `Solver_env.t` so I'll leave it for now.

gridbugs commented:
@rgrinberg does anything more need to happen before this can be merged?

rgrinberg commented:
Ready from me. Waiting on @Leonidas-from-XIV to review

Leonidas-from-XIV commented on test/blackbox-tests/test-cases/pkg/solver-flags-per-context.t:
```diff
@@ -0,0 +1,102 @@
+  >   (name with-doc-and-with-test)
+  >   (lock with-doc-and-with-test.lock)
+  >   (solver_env
+  >    (flags with-doc with-test))))
```
I wonder whether it would make sense to implement the set language so `(flags :standard)` can be specified.

gridbugs replied:
Ah yes I like the idea of allowing `:standard` here. It won't be hard to add as long as we don't allow `:include` which I think is reasonable.

gridbugs replied:
Updated to use the ordered set language for this field and added a test using the `:standard` keyword.

Leonidas-from-XIV reviewed (approved):
I think it looks quite good, some minor comments when I looked over the diff.

gridbugs commented:
> Yes, that's what I meant. Otherwise, we'll have to always build all the test deps.

@rgrinberg on the topic of storing additional metadata to keep track of which dependencies are test/doc-only so we don't build test/doc deps when installing, I thought about this some more and I don't think this is necessary. The solver flags apply only to dependencies of local packages define in `dune-project` and the dependency specification in `dune-project` already stores whether a package is a test or doc dependency.

rgrinberg committed:
chore: fix sexp formatting

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>
